File: frontend/src/lib/rows.ts
/** The form for writing one row, derived from the table's own columns.
 *
 *  Every question this file answers is a question about a *declared type*, not
 *  about a value: which control a column gets, whether it may be left empty,
 *  what an enum's members are. That is the same discipline the chart suggester
 *  follows and for the same reason — a 64-bit integer arrives as a string, so
 *  the type is the only thing that can be reasoned about.
 *
 *  Nothing here formats a value into SQL. The strings this produces travel to
 *  the server as query parameters declared with the column's own type, and the
 *  server parses each one against the type it is about to store it in. So this
 *  file's job is to ask the right question, never to validate the answer: a
 *  browser-side check of what fits in a `Decimal(38,10)` would be a second,
 *  worse implementation of something ClickHouse already does exactly. */

import { unwrap } from './chType'

/** What a column needs from the person filling in the form. */
export type Control =
  /** A fixed set, from the type itself. */
  | 'enum'
  /** Two states and nothing else. */
  | 'bool'
  /** A single line. Everything that is not one of the above. */
  | 'text'
  /** Many lines: the types where a real value is routinely long. */
  | 'long'

export interface Column {
  name: string
  type: string
  default_kind: string
  default_expression: string
  comment: string
  nullable: boolean
}

export interface Field {
  column: Column
  control: Control
  /** An enum's members, in the order the type declares them. Empty otherwise. */
  members: string[]
  /** Whether leaving this alone is a real option — either the table computes a
   *  value for it or it accepts a null. */
  optional: boolean
  /** What happens if it is left alone, in words, or null where nothing can be
   *  said with certainty. */
  ifLeftAlone: string | null
}

/** Whether a value may be written into this column at all.
 *
 *  `MATERIALIZED` and `ALIAS` columns are there and are computed from the
 *  others, which is a different fact from being absent — and the server's two
 *  refusals say neither. Naming a materialized column in an insert answers
 *  *Cannot insert column c, because it is MATERIALIZED column*; naming an alias
 *  answers *No such column d in table*, as though it were not there at all.
 *  So the form leaves both out and says why, rather than offering a box that
 *  cannot work. */
export function writable(column: Column): boolean {
  return column.default_kind !== 'MATERIALIZED' && column.default_kind !== 'ALIAS'
}

/** An `Enum8('ok' = 1, 'bad' = 2)`'s members, in declaration order.
 *
 *  Read off the type because that is where they are: ClickHouse writes the
 *  whole set into `system.columns.type`, so a list needs no second query and
 *  cannot drift from the column it belongs to.
 *
 *  The quoting is the awkward part and it is real: a member may contain a
 *  comma, a bracket, or an escaped quote — `Enum8('a,b' = 1, 'it\'s' = 2)` is a
 *  legal type — so the members are matched as quoted strings rather than split
 *  on commas. */
export function enumMembers(type: string): string[] {
  const inner = /^Enum(?:8|16)?\((.*)\)$/s.exec(unwrap(type))
  if (!inner?.[1]) return []
  const out: string[] = []
  const re = /'((?:[^'\\]|\\.)*)'/g
  let m: RegExpExecArray | null
  while ((m = re.exec(inner[1])) !== null) {
    out.push((m[1] ?? '').replace(/\\'/g, "'").replace(/\\\\/g, '\\'))
  }
  return out
}

/** The types where a real value is routinely longer than a line.
 *
 *  `String` is deliberately not among them, and that was measured by looking at
 *  a form rather than reasoned: it is ClickHouse's *only* string type, so it
 *  holds an email address as often as it holds a JSON blob, and giving every
 *  one of them a textarea turned an eight-column table into a page of boxes.
 *  What is left are the types that are structured by definition — a `Map` or a
 *  `Tuple` written on one line is unreadable however short it is. */
const LONG = /^(JSON|Object|Array|Map|Tuple|Nested|Variant|Dynamic)/

/** Which control a column gets. */
export function controlFor(type: string): Control {
  const bare = unwrap(type)
  if (/^Enum(8|16)?\(/.test(bare)) return 'enum'
  if (/^Bool$/.test(bare)) return 'bool'
  if (LONG.test(bare)) return 'long'
  return 'text'
}

/** Whether the declared type accepts a null.
 *
 *  Read off the type rather than trusted from `ColumnDetail.nullable`, so the
 *  answer is right for a `LowCardinality(Nullable(String))` — where the
 *  wrapper is on the inside and a naive prefix test says no. */
export function nullable(type: string): boolean {
  return /\bNullable\(/.test(type)
}

/** The form, in the order the table declares its columns.
 *
 *  Declaration order and not, say, keys first: this is the shape somebody
 *  already has in their head from the Columns tab, and re-ordering the same
 *  list differently in two places is how a reader loses track of which table
 *  they are looking at. */
export function fieldsFor(columns: Column[]): Field[] {
  return columns.filter(writable).map((column) => {
    const isNull = nullable(column.type)
    const hasDefault = column.default_kind === 'DEFAULT'
    return {
      column,
      control: controlFor(column.type),
      members: enumMembers(column.type),
      optional: isNull || hasDefault,
      ifLeftAlone: hasDefault
        ? `the table writes ${column.default_expression}`
        : isNull
          ? 'the row gets a null'
          : /* Not "the type's zero", which is what ClickHouse will in fact
               write: `''` for a String, `0` for a UInt8, the epoch for a
               DateTime. Saying so would be true and would read as a promise
               Flint is making, when it is the engine's behaviour and differs
               per type. The form asks for a value instead. */
            null,
    }
  })
}

/** The columns the form is not offering, and why — one sentence, or null where
 *  there are none.
 *
 *  Said rather than silently dropped, on the same rule as every other fold in
 *  the product: a form showing nine of a table's eleven columns, with nothing
 *  saying so, reads as a form that has lost two. */
export function saysComputed(columns: Column[]): string | null {
  const computed = columns.filter((c) => !writable(c))
  if (computed.length === 0) return null
  const names = computed.map((c) => `\`${c.name}\``).join(', ')
  return `${names} ${computed.length === 1 ? 'is' : 'are'} computed by the server from the other columns, so ${
    computed.length === 1 ? 'it is' : 'they are'
  } not written with the row.`
}

/** What a filled-in form sends.
 *
 *  Three states per column, and they have to stay three: a value, an explicit
 *  null, and *not mentioned at all*. The last is what makes a `DEFAULT` apply,
 *  and it is why "leave alone" cannot be a magic string — any string that meant
 *  *default* would be a string somebody might have meant to store.
 *
 *  And an empty box is an empty string, not a null. Measured against a server
 *  rather than assumed: binding `''` to a `Nullable(String)` stores a
 *  zero-length string and `IS NULL` comes back false. Two different answers, so
 *  the form has two different ways to ask for them. */
export type Entry = { kind: 'value'; text: string } | { kind: 'null' } | { kind: 'default' }

export interface Payload {
  column: string
  value: string | null
}

export function payload(fields: Field[], entries: Record<string, Entry>): Payload[] {
  const out: Payload[] = []
  for (const field of fields) {
    const entry = entries[field.column.name] ?? { kind: 'default' }
    if (entry.kind === 'default') continue
    out.push({ column: field.column.name, value: entry.kind === 'null' ? null : entry.text })
  }
  return out
}

/** What the table will fill in, said before the button rather than reported
 *  after it. A row that comes back holding three columns nobody typed is a row
 *  somebody wanted to have been warned about first.
 *
 *  A string with backticks in it, rendered by `Sentence` — which is the
 *  convention the whole codebase writes in, and the one that keeps the wording
 *  assertable in a test. Building the list as JSX instead put literal
 *  backticks on the screen, which by `Note.tsx`'s own account is now the third
 *  time that has shipped. */
export function saysDefaulting(fields: Field[], entries: Record<string, Entry>): string | null {
  const left = willDefault(fields, entries)
  if (left.length === 0) return null
  const names = left.map((n) => `\`${n}\``).join(', ')
  return `${left.length} of ${fields.length} ${
    left.length === 1 ? 'column is' : 'columns are'
  } left to the table: ${names}.`
}

/** What the server reported after the write. Its answer and not the form's:
 *  the route read the table's own columns at the moment of writing, and that
 *  is the list that is actually true of the row now on disk. */
export function saysWritten(defaulted: string[]): string {
  if (defaulted.length === 0) return 'Written.'
  const names = defaulted.map((n) => `\`${n}\``).join(', ')
  return `Written. The table filled in ${names}.`
}

export function willDefault(fields: Field[], entries: Record<string, Entry>): string[] {
  return fields
    .filter((f) => (entries[f.column.name] ?? { kind: 'default' }).kind === 'default')
    .map((f) => f.column.name)
}

File: src/routes/rows.rs
//! Writing rows: the Data side of A1.
//!
//! Rows are Data and structure is Infrastructure, and that line is why this
//! file exists rather than the write being folded into `jobs.rs` beside the
//! `ALTER`s. Adding a row changes nothing about what the table *is*, so it
//! belongs on the table's own page, where the person reading the rows already
//! is — and the two-space rule holds precisely because nothing here can change
//! a column, a key or an engine.
//!
//! It is deliberately synchronous. Every other write Flint makes is a job,
//! because every other write is an `OPTIMIZE`, a `BACKUP` or a mutation that
//! can run for an hour; a single-row insert answers in milliseconds and a job
//! row for it would be a receipt nobody wanted, in a list whose job is to show
//! the things still running.

use std::collections::HashSet;

use axum::{extract::State, Json};
use serde::{Deserialize, Serialize};

use super::{AppState, Caller};
use crate::clickhouse::rows::{self, Column, Given};
use crate::clickhouse::QueryOptions;
use crate::config::Tier;
use crate::error::{Error, Result};

#[derive(Debug, Deserialize)]
pub struct InsertRequest {
    pub database: String,
    pub table: String,
    /// One entry per column being written. A column that is not here is left
    /// out of the statement, which is how its `DEFAULT` applies.
    pub fields: Vec<Field>,
}

#[derive(Debug, Deserialize)]
pub struct Field {
    pub column: String,
    /// `null` is SQL `NULL`. A string is a string — including `""`, which is a
    /// zero-length string and not a null, because the server treats them as
    /// two different answers and so does the form.
    pub value: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct Written {
    /// What ran, verbatim. "What did that button actually run" is the first
    /// question anybody asks of a tool that writes on their behalf, and the
    /// bound values are deliberately *not* spliced into it — the statement is
    /// the shape, the values were the parameters.
    pub statement: String,
    /// The columns the table filled in itself, because they were left out.
    /// Named rather than counted: a row that came back with three columns you
    /// did not type is a row you want to be told about.
    pub defaulted: Vec<String>,
}

/// `POST /api/rows` — one row into one table.
pub async fn insert(
    State(state): State<AppState>,
    Caller(ch): Caller,
    Json(req): Json<InsertRequest>,
) -> Result<Json<Written>> {
    state.require_tier(Tier::Data)?;
    let database = name(&req.database, "database")?;
    let table = name(&req.table, "table")?;

    // The table's own columns, from the server, now. Both the names and the
    // declared types go into the statement — one as an identifier, one inside a
    // parameter declaration — and neither is a literal that could be bound, so
    // neither may come from the caller.
    let columns = rows::columns(&ch, &database, &table).await?;

    let mut chosen: Vec<(&Column, Given)> = Vec::with_capacity(req.fields.len());
    let mut seen: HashSet<&str> = HashSet::new();
    for field in &req.fields {
        let Some(column) = columns.iter().find(|c| c.name == field.column) else {
            return Err(Error::BadRequest(format!(
                "`{}` has no column called `{}`",
                qualified(&database, &table),
                field.column
            )));
        };
        if !seen.insert(column.name.as_str()) {
            return Err(Error::BadRequest(format!(
                "`{}` was given twice, and a row has one value per column",
                column.name
            )));
        }
        // Said in Flint's words because the server's two refusals here are
        // different codes and neither says what is actually true: the column is
        // there, and it is computed from the others rather than written.
        if !column.writable() {
            return Err(Error::BadRequest(format!(
                "`{}` is {} — the server computes it from `{}`, so it is not \
                 written with the row",
                column.name,
                column.default_kind.to_lowercase(),
                column.default_expression
            )));
        }
        chosen.push((
            column,
            match &field.value {
                None => Given::Null,
                Some(text) => Given::Value(text.clone()),
            },
        ));
    }

    if chosen.is_empty() {
        // `INSERT INTO t () VALUES ()` is not a statement, and "a row of
        // nothing but defaults" is a thing somebody could mean but never a
        // thing they meant by leaving every box alone.
        return Err(Error::BadRequest(
            "nothing was filled in, so there is no row to write".into(),
        ));
    }

    let statement = rows::insert(&database, &table, &chosen);
    ch.execute(
        &statement.sql,
        QueryOptions {
            params: statement.params.clone(),
            ..Default::default()
        },
    )
    .await?;

    // Every writable column nobody filled in. The unwritable ones are left out
    // of this list on purpose: they were never on offer, so reporting them as
    // "defaulted" would suggest a choice was made about them.
    let given: HashSet<&str> = chosen.iter().map(|(c, _)| c.name.as_str()).collect();
    let defaulted = columns
        .iter()
        .filter(|c| c.writable() && !given.contains(c.name.as_str()))
        .map(|c| c.name.clone())
        .collect();

    Ok(Json(Written {
        statement: statement.sql,
        defaulted,
    }))
}

/// `GET /api/rows/columns` is not a route: the table page already holds every
/// column's name, type, default and comment from `table_detail`, and asking a
/// second time would let the form drift from the page it opened on.
fn qualified(database: &str, table: &str) -> String {
    format!("{database}.{table}")
}

/// The same shape-check `jobs.rs` applies, for the same reason: the identifier
/// is escaped when the statement is built, so what is refused here is the shape
/// that has no business being a name at all.
fn name(raw: &str, what: &str) -> Result<String> {
    let trimmed = raw.trim();
    if trimmed.is_empty() {
        return Err(Error::BadRequest(format!("a {what} name is required")));
    }
    if trimmed.len() > 255 {
        return Err(Error::BadRequest(format!(
            "that {what} name is longer than any ClickHouse identifier"
        )));
    }
    if trimmed.chars().any(|c| c.is_control()) {
        return Err(Error::BadRequest(format!(
            "that {what} name contains a control character"
        )));
    }
    Ok(trimmed.to_string())
}
